Add Polkit Quick Unlock mechanism

HexF opened the pull request:
[NOTE]: # ( Describe your changes in detail, why is this change required? )

This pull request closes 2 issues relating to the Quick Unlock feature (commonly referred to as fingerprint support) - #5991 and #3337.
The changes made implement quick unlocking via [Polkit](https://gitlab.freedesktop.org/polkit/polkit/) on Linux, which provides support for fingerprint authentication, giving Linux feature parity with macOS with its TouchID and Windows with Windows Hello.
Additionally, polkit will fall back to a password if no fingerprint sensor is found, as shown in the screenshots below. 

The Polkit mechanism works similar to the existing TouchID implementation, by where the database key is encrypted using a random key which is stored in secure storage, in this case the [kernel-space keyring](https://man7.org/linux/man-pages/man2/add_key.2.html).
As far as I am aware, Linux does not provide a kernel-level mechanism for releasing keys when appropriate credentials such as a password or fingerprint have been provided, so the check is done entirely in the KeepassXC process.

This pull request also makes changes to both the TouchID and Windows Hello implementations of Quick Unlocking, bringing them on to a common interface which should allow for easy development of further Quick Unlock mechanisms - one that comes to mind is using TOTP/one time passwords to unlock the database (#798). The polkit implementation also uses this interface.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

![image](https://user-images.githubusercontent.com/11236933/212122322-b601cb73-d4cb-407a-88b5-fede8ca8d179.png)
![image](https://user-images.githubusercontent.com/11236933/212122376-d10a9daf-a9a1-4841-a9ea-89fdae76dfd6.png)
![image](https://user-images.githubusercontent.com/11236933/212122693-de73d221-b149-4ff2-9bcf-f71d1b25dd9e.png)
(please excuse the polkit action having the wrong wording in the above screenshot - it is surprisingly hard to add polkit actions on nixos without installing anything!)

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Local testing has been carried out through the following steps:
1. Open and unlock a database
2. Lock the database
3. Observe the Quick Unlock "Unlock Database" button being present
4. Click the button
5. Observe a polkit authentication request
6. Satisfy this request through entering a password
7. Observe the Keepass database unlocked

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (change that adds functionality)
- ✅ Refactor (significant modification to existing code)

HexF commented:
Not sure how to go about adding the `libkeyutils-dev` dependency to all the CI which seems to be whats failing on Linux.

droidmonkey commented:
That's on our end we'll take care of that

> This pull request also makes changes to both the TouchID and Windows Hello implementations of Quick Unlocking, bringing them on to a common interface which should allow for easy development of further Quick Unlock mechanisms

You are my hero thank you

michaelk83 commented on share/linux/org.keepassxc.KeePassXC.policy.in:
```diff
@@ -0,0 +1,18 @@
+      <allow_active>auth_self_one_shot</allow_active>
+    </defaults>
+  </action>
+</policyconfig>
```
Should this not have a newline here?

michaelk83 commented on src/gui/reports/ReportsDialog.cpp:
```diff
@@ -31,7 +31,7 @@
 #include "core/Group.h"
 #ifdef Q_OS_MACOS
-#include "touchid/TouchID.h"
+#include "quickunlock/TouchID.h"
```
Where is this used here, and why only TouchID?
(probably can be removed)

HexF replied:
Yeah I just went on a find/replace for all `touchid/` to `quickunlock/`, probably can be removed if not used elsewhere in the file

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -0,0 +1,192 @@
+QString getKeyName(const QString& dbPath)
+{
+    static const QString keyPrefix = "keepassxc_polkit_keys_";
+    const QByteArray pathHash = CryptoHash::hash(dbPath.toUtf8(), CryptoHash::Sha256).toHex();
```
There was some discussion of using UUIDs in https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-871298047 and below (search for "uuid").

HexF replied:
This has been completed, using a new method `Database->publicUuid()`.

michaelk83 commented on src/quickunlock/Polkit.h:
```diff
@@ -0,0 +1,34 @@
+    bool hasKey(const QString& dbPath) const;
+    void reset(const QString& dbPath);
+
+    Polkit();
```
Does KPXC have code guidelines for method order? In my own code I usually place the constructor and destructor first, then order the rest alphabetically within each visibility group (with the exception of pairing matching getters and setters together). Same in the implementation.

In particular, I'm bothered by the two `reset()` overloads not being next to each other. Maybe the 2nd one should be `remove()`? Consider also renaming the first one to `clear()` and "store" to "set".

droidmonkey replied:
Yes, we do have code ordering guidelines. They are in the CONTRIBUTING document.

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -0,0 +1,192 @@
+    Botan::secure_scrub_memory(keychainKeyValue.data(), keychainKeyValue.size());
+
+    // Store encrypted master password and return
+    m_encryptedMasterKeys.insert(dbPath, encryptedMasterKey);
```
Would it make any more sense to store the encrypted master key in kernel memory (in the keyring), and the `keychainKeyValue` in process memory?

Also, if you're scrubbing the other local variables, why not `encryptedMasterKey` (after this line)?

HexF replied:
I did it the same way as in TouchID, and this is the best way seeing that *in theory* someone could dump the value from the Linux keyring, so requiring both the process to hold this in the memory **and** in the Linux keyring is for the better.
Besides, it adds what - <1KB of memory usage total and a few milliseconds of time to unlock the database, for a slight security benefit.

As for encryptedMasterKey, its being stored in a hash so why scrub it?

michaelk83 replied:
Kernel memory is supposed to be protected. AFAIK, KPXC takes measures to protect its process memory too, but it's still userspace. I don't actually know what difference there is in practice though, if any.

My reasoning both for moving `encryptedMasterKey` to kernel memory and scrubbing it from the process memory is based on the fact that you need three pieces of information to unlock the database: the encrypted master key, the encryption key, and the encryption IV. If you get all three, you can obtain the unencrypted master key, and gain access to the database.

All three of them are sensitive, in a way. `encryptedMasterKey` is a symmetric encryption of the master key, not a unidirectional hash.

So then:
A. You want to keep as few copies as possible of *all* the sensitive data.
B. The next question is, what's the safest way to "spread your eggs"?

My instinct was that the mater key is more sensitive, even when encrypted. But on 2nd thought, I guess there isn't really a difference, as long as they're not all stored in the same place. Especially given that the whole database is stored in process memory when unencrypted, anyway.

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -0,0 +1,192 @@
+    QMap<QString, QString> details;
+
+    auto result = m_polkit->CheckAuthorization(subject,
+                                               // "org.keepassxc.KeePassXC.unlockDatabase",
```
I don't see the `org.keepassxc.KeePassXC.unlockDatabase` action used anywhere except here. Is it still needed?

HexF replied:
Yeah I forgot to fix that while testing - didn't exactly use my own action. I'll get that fixed.

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        // Scrub the keys from ram
+        Botan::secure_scrub_memory(keychainKey.data(), keychainKey.size());
+        Botan::secure_scrub_memory(keychainIv.data(), keychainIv.size());
+        Botan::secure_scrub_memory(keychainBytes.data(), keychainBytes.size());
```
Why not `encryptedMasterKey`?

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -0,0 +1,192 @@
+            return false;
+        }
+
+        void* keychainBuffer;
+        long keychainDataSize = keyctl_read_alloc(keySerial, &keychainBuffer);
```
Does this get freed automatically somewhere? Otherwise you have a memory leak.

HexF replied:
Good catch

michaelk83 commented on src/quickunlock/PolkitDbusTypes.h:
```diff
@@ -0,0 +1,36 @@
+Q_DECLARE_METATYPE(PolkitSubject);
+Q_DECLARE_METATYPE(PolkitAuthorizationResults);
+
+#endif // KEEPASSX_POLKITDBUSTYPES_H
```
newline

michaelk83 commented on src/quickunlock/QuickUnlockInterface.cpp:
```diff
@@ -0,0 +1,25 @@
+        quickUnlockInstance = new QUICKUNLOCK_IMPLEMENTATION();
+    }
+    return quickUnlockInstance;
+}
```
newline

michaelk83 commented on src/quickunlock/QuickUnlockInterface.h:
```diff
@@ -0,0 +1,67 @@
+
+    QString errorString() const
+    {
+        return "No Quick Unlock provider is available";
```
Translations.
Is "Quick Unlock" with or without a space elsewhere?

droidmonkey replied:
Also don't define functions in the header file, especially those that return non trivial values.

michaelk83 commented on src/quickunlock/WindowsHello.cpp:
```diff
@@ -97,23 +97,10 @@ namespace
-        m_available = state;
-        emit availableChanged(m_available);
-    });
+    m_available = KeyCredentialManager::IsSupportedAsync().get();
```
This is apparently an asynchronous call, so this change may not work.

HexF replied:
Would be ideal if this was sync - if someone can test on Windows that would be great

michaelk83 replied:
I [looked it up](https://learn.microsoft.com/en-us/uwp/api/windows.security.credentials.keycredentialmanager.issupportedasync?view=winrt-22621#windows-security-credentials-keycredentialmanager-issupportedasync) and initially I didn't see how the asynchronicity works here, since there's apparently no callback. But after digging a little deeper, I see the `.get()` call [makes it synchronous](https://learn.microsoft.com/en-us/uwp/api/windows.foundation.iasyncoperation-1?view=winrt-22621#cwinrt-extension-functions). So it should work.

Might want to add a try-catch for it, though.

droidmonkey replied:
Why was windowshello code touched at all?

michaelk83 replied:
I suppose to get it in line with the new common interface. The previous code was using a signal that's not part of the interface.

Technically, this could just skip the `emit` call, but after a skim of the API docs, I think the new code is enough.

droidmonkey replied:
The .get() could cause the ui to freeze if windows takes a while to process the request.

michaelk83 commented on src/quickunlock/dbus/org.freedesktop.PolicyKit1.Authority.xml:
```diff
@@ -0,0 +1,16 @@
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In2" value="QMap&lt;QString, QString&gt;"/>
+    </method>
+  </interface>
+</node>
```
newline?

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -0,0 +1,192 @@
+                                               // "org.keepassxc.KeePassXC.unlockDatabase",
+                                               "org.freedesktop.policykit.exec",
+                                               details,
+                                               0x00000001,
```
What does this flag do? (Add a comment)

klemensn commented on src/CMakeLists.txt:
```diff
@@ -206,6 +207,8 @@ endif()
 if(UNIX AND NOT APPLE)
```
Can you wrap your new code in a new `if(LINUX)` under this?
Otherwise you'll also enable it for all BSDs on which it probably wasn't tested so far.

klemensn commented on share/CMakeLists.txt:
```diff
@@ -58,7 +58,10 @@ if(UNIX AND NOT APPLE AND NOT HAIKU)
+    configure_file(linux/${APP_ID}.policy.in ${CMAKE_CURRENT_BINARY_DIR}/linux/${APP_ID}.policy @ONLY)
+
     install(FILES ${CMAKE_CURRENT_BINARY_DIR}/linux/${APP_ID}.desktop DESTINATION ${CMAKE_INSTALL_DATADIR}/applications)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/linux/${APP_ID}.policy DESTINATION ${CMAKE_INSTALL_DATADIR}/polkit-1/actions)
```
The XML file is not a problem, but like the actual code this block here is under `if(UNIX AND NOT APPLE AND NOT HAIKU)`.

Can you wrap it in `if(LINUX)` to exclude BSDs here?

klemensn commented on src/CMakeLists.txt:
```diff
@@ -207,10 +207,14 @@ endif()
-            quickunlock/PolkitDbusTypes.cpp
             gui/osutils/nixutils/ScreenLockListenerDBus.cpp
             gui/osutils/nixutils/NixUtils.cpp)
+    if("${CMAKE_SYSTEM}" MATCHES "Linux")
```
I suggest sticking to `if(LINUX)` as done in previous versions of this CMake code.

It is consistent with `if(... UNIX ... APPLE ... HAIKU ...)` and that one
ne occurence in src/CMakeLists.txt currently using `if("${CMAKE_SYSTEM}" STREQUAL "Linux")` slipped in with some macOS change, probably by accident, replacing the previous `if(LINUX)`.

HexF replied:
`if(LINUX)` didn't work on my system but that did, and im unsure as to why, but that did

HexF replied:
I can't find reference to `if(LINUX)` being valid anywhere in the [CMake docs](https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/How-To-Write-Platform-Checks#platform-variables).

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -82,18 +83,37 @@ bool Polkit::getKey(const QString& dbPath, QByteArray& key)
 
+    pid_t app_pid = QCoreApplication::applicationPid();
+
+    QString processStatPath = QString("/proc/%1/stat").arg(app_pid);
+    QFile processStatFile(processStatPath);
+
+    if(!processStatFile.open(QIODevice::ReadOnly | QIODevice::Text)){
+        qDebug() << "polkit: failed to open " << processStatPath << " for reading";
+        return false;
+    }
+
+    QTextStream processStatStream(&processStatFile);
+    QString processStatInfo = processStatStream.readLine();
+    processStatFile.close();
+
+    QString processStartTime = processStatInfo.split(" ")[21];
```
May want to move this to a helper function in [`gui/osutils/nixutils`](https://github.com/keepassxreboot/keepassxc/tree/develop/src/gui/osutils/nixutils).

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -82,18 +83,37 @@ bool Polkit::getKey(const QString& dbPath, QByteArray& key)
+                                               "org.keepassxc.KeePassXC.unlockDatabase",
                                                details,
-                                               0x00000001,
+                                               0x00000001, 
+                                               // AllowUserInteraction - wait for user to authenticate
```
If there's a usable header for this, it'd be more readable to just use the enum constant. But this is ok too.

HexF replied:
There are no header libraries available that I can find.

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -128,12 +148,16 @@ bool Polkit::getKey(const QString& dbPath, QByteArray& key)
 
         QByteArray keychainBytes(static_cast<const char*>(keychainBuffer), keychainDataSize);
+
+        free(keychainBuffer);
```
Since you're scrubbing `keychainBytes`, you should also scrub `keychainBuffer` before freeing it.

michaelk83 commented on src/quickunlock/Polkit.cpp:
```diff
@@ -149,6 +173,7 @@ bool Polkit::getKey(const QString& dbPath, QByteArray& key)
         Botan::secure_scrub_memory(keychainKey.data(), keychainKey.size());
         Botan::secure_scrub_memory(keychainIv.data(), keychainIv.size());
         Botan::secure_scrub_memory(keychainBytes.data(), keychainBytes.size());
+        Botan::secure_scrub_memory(encryptedMasterKey.data(), encryptedMasterKey.size());
```
Hmm.. should these also be scrubbed when removed from `m_encryptedMasterKeys`? @droidmonkey , what's your opinion - is this overkill?

droidmonkey replied:
Anything encrypted in memory is already "random" data so it doesn't matter.

HexF commented:
Haven't touched this in a bit, but now that the vulnerability in 1Password that I discovered has been fixed I feel it's appropriate to discuss it here as it has direct implications on the use of PolKit.

The vulnerability boils down to being able to redirect D-Bus calls from the password manager to polkit into your own service. This happens by adjusting the DBUS_SYSTEM_BUS_ADDRESS environment variable when the application is started.

I didn't investigate the exact way 1Password fixed this, but there are two ways which I foresee fixing this for KeePassXC

1. Run a separate process as root which manages the releasing of partial key data from PolKit when authorization is completed
2. Clear environment variables, ignoring the changes to the system bus address (pkexec does this)

I'm leaning more towards a separate process, but there is a large inconvenience of having to install a system service for this.

Any opinions @droidmonkey

HexF commented:
Haven't touched this in a bit, but now that the vulnerability in 1Password that I discovered has been fixed I feel it's appropriate to discuss it here as it has direct implications on the use of PolKit.

The vulnerability boils down to being able to redirect D-Bus calls from the password manager to polkit into your own service. This happens by adjusting the DBUS_SYSTEM_BUS_ADDRESS environment variable when the application is started.

I didn't investigate the exact way 1Password fixed this, but there are two ways which I foresee fixing this for KeePassXC

1. Run a separate process as root which manages the releasing of partial key data from PolKit when authorization is completed
2. Clear environment variables, ignoring the changes to the system bus address (pkexec does this)

I'm leaning more towards a separate process, but there is a really large inconvenience

droidmonkey commented:
We won't install a system service on our end. This boils down to a Linux environment vulnerability not an application vulnerability. Choice 1 is something either polkit or the desktop env should implement for the benefit of everyone. Choice 2 is a workaround because of a lack of choice 1.

Let's just capture the value of the environment variable on startup and/or ignore the actual value.

Gimli05 commented:
Hi!

Can we get a status update on this one? 

Is polkit excluded due to the dbus redirect issue? Saving the env var value on startup and ignoring its actual value during runtime seems to be a valid option to me. However I'm not sure whether storing it for just this purpose would be a nice solution, rather than a hacky one. On the other hand I cannot see a valid purpose why the dbus address would need to be changed during runtime.

HexF commented:
Hi, Sorry been really busy.

I am thinking as a way to combat the environment variable making it easy to set, we compile in a list of allowed paths for the dbus system bus.
If it does not match any of the approved paths, QuickUnlock will be disabled and/or display a nag bar indicating the presence of an untrustworthy path.

michaelk83 commented:
A hardcoded path whitelist sounds good to me. If the path doesn't match, I'm in favor of *both* disabling QuickUnlock and showing a warning similar to what the Secret Service component does when there's another Secret Service provider (such as Gnome keyring) already running.
https://github.com/keepassxreboot/keepassxc/blob/bb37cf3c3235b612d737725e5523e41796f7126d/src/fdosecrets/dbus/DBusMgr.cpp#L412

HexF commented:
New changes implement a single fixed path to access DBus from, which should contain polkit.
The only vulnerability I can see here is if polkit is not installed on the system, and then a malicious actor simply claims the service on the dbus.
I can't see any way around this, and the chances of it occurring are very slim, as most modern Linux distros run Polkit.

Changes also implement a `publicUuid`, which is stored in the pubilc headers headers as `KPXC_PUBLIC_UUID`.
This is used instead of the file path to reference the database for quick unlocking.
`uuid` was insufficient as it was randomized every time the database was locked/unlocked.

I'll try get to these merge conflicts tomorrow, but in the mean time any feedback is much appreciated. 
I assume these CI failures are due to a fairly outdated repo.

HexF commented:
Rebased to upstream/develop

HexF commented:
At this point in time, I feel the code is functionality-wise, everything that will make it in with this PR.
Before proceeding with merging, I would like manual testing on both macOS and Windows to be carried out for the quick unlock functionality as this PR could have possibly broken that.

As for the CI failure on Linux - the additional dependency for `keyutils` needs to be added to fix that.

PR is ready for a functionality and code review.

droidmonkey commented:
Excellent will review. @phoerious can you test macOS, I can never get my signing and notorizing to work...

droidmonkey commented:
An interesting consequence of injecting the Public UUID is that the database will be marked modified upon unlock. If the user doesn't save the database (if auto-save is turned off), then quick unlock remains disabled. This creates a rather challenging UX problem.

Also I feel dumb because I can't get this to build on WSL:

```
FAILED: src/keepassxc 
: && /usr/bin/c++ -fno-common -fopenmp -Wall -Wextra -Wundef -Wpointer-arith -Wno-long-long -Wformat=2 -Wmissing-format-attribute -fvisibility=hidden -fvisibility-inlines-hidden -Wshadow-compatible-local -Wshadow-local -Werror -fstack-protector-strong -Wnon-virtual-dtor -Wold-style-cast -Woverloaded-virtual -fsanitize=address -DWITH_ASAN -fsanitize=leak -DWITH_LSAN -Werror=format-security -Wcast-align -fsized-deallocation -Wno-deprecated-declarations -g -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro,-z,now -pie   -Wl,--export-dynamic -rdynamic  -fPIE -pie src/CMakeFiles/keepassxc.dir/keepassxc_autogen/mocs_compilation.cpp.o src/CMakeFiles/keepassxc.dir/main.cpp.o -o src/keepassxc  src/libkeepassx_core.a  src/libautotype.a  src/browser/libkeepassxcbrowser.a  src/qrcode/libqrcode.a  /usr/lib/x86_64-linux-gnu/libQt5Svg.so.5.15.3  /usr/lib/x86_64-linux-gnu/libqrencode.so  src/fdosecrets/libfdosecrets.a  -lkeyutils  /usr/lib/x86_64-linux-gnu/libQt5Concurrent.so.5.15.3  -lpcsclite  src/libzxcvbn.a  /usr/lib/x86_64-linux-gnu/libargon2.so  src/thirdparty/ykcore/libykcore.a  /usr/lib/x86_64-linux-gnu/libusb-1.0.so  src/sshagent/libsshagent.a  /usr/lib/x86_64-linux-gnu/libQt5Network.so.5.15.3  src/keeshare/libkeeshare.a  /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5.15.3  /usr/lib/x86_64-linux-gnu/libbotan-2.so  /usr/lib/x86_64-linux-gnu/libz.so  /usr/lib/x86_64-linux-gnu/libminizip.so  /usr/lib/x86_64-linux-gnu/libQt5DBus.so.5.15.3  /usr/lib/x86_64-linux-gnu/libQt5X11Extras.so.5.15.3  /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.15.3  /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.15.3  -lX11 && :
/usr/bin/ld: src/libkeepassx_core.a(Polkit.cpp.o): in function `Polkit::setKey(QUuid const&, QByteArray const&)':
/home/jon/.vs/keepassxc/src/quickunlock/Polkit.cpp:79: undefined reference to `add_key(char const*, char const*, void const*, unsigned long, int)'
/usr/bin/ld: src/libkeepassx_core.a(Polkit.cpp.o): in function `Polkit::getKey(QUuid const&, QByteArray&)':
/home/jon/.vs/keepassxc/src/quickunlock/Polkit.cpp:139: undefined reference to `find_key_by_type_and_desc(char const*, char const*, int)'
/usr/bin/ld: /home/jon/.vs/keepassxc/src/quickunlock/Polkit.cpp:148: undefined reference to `keyctl_read_alloc(int, void**)'
/usr/bin/ld: src/libkeepassx_core.a(Polkit.cpp.o): in function `Polkit::hasKey(QUuid const&) const':
/home/jon/.vs/keepassxc/src/quickunlock/Polkit.cpp:202: undefined reference to `find_key_by_type_and_desc(char const*, char const*, int)'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

Perhaps WSL doesn't support polkit at all??

HexF commented:
Those errors dont look like Polkit errors, rather they look like `keyutils` isn't linking properly on WSL.

Functions `find_key_by_type_and_desc` and `add_key` are part of `keyutils`. 
My thinking is that WSL doesn't implement the keyring service provided by the Linux kernel.

droidmonkey commented:
I figured it out....

keyutils version 1.6.1 is completely broken. Not only does it not wrap functions with `extern "C"` but it also uses the keyword `private` in a function parameter. Fedora 37 and 38 deploys 1.6.1 and so does Ubuntu 22.04.

https://salsa.debian.org/debian/keyutils/-/commit/2285153c6992d76e649808fe87f9bafc327b0d1e#8f6252b6c15417d4eeec4c552fa0711c0d936bcb_18_19

https://salsa.debian.org/debian/keyutils/-/commit/2285153c6992d76e649808fe87f9bafc327b0d1e#8f6252b6c15417d4eeec4c552fa0711c0d936bcb_222_229

HexF commented:
Any updates on getting the CI to pass?
Seems the CI doesn't have keyutils installed at all.

droidmonkey commented:
I added it to the docker file but it doesn't seem like it took.

phoerious commented:
Should work now. Failures are now due to code errors:

```
/opt/buildagent/work/c401303cba1b4098/src/quickunlock/Polkit.cpp: In function ‘QString getKeyName(const QUuid&)’:
/opt/buildagent/work/c401303cba1b4098/src/quickunlock/Polkit.cpp:40:47: error: ‘Id128’ is not a member of ‘QUuid’
       return keyPrefix + dbUuid.toString(QUuid::Id128);
```

HexF commented:
`QUuid::Id128` appears to have been added in Qt 5.11 - any chance we could upgrade the CI to that version, and hence raise the minimum required Qt version?

droidmonkey commented:
You need to use a supported api

droidmonkey commented:
Well have to update to qt6 first

phoerious commented:
We cannot do that without sacrificing Ubuntu LTS compatibility. Our minimum Qt version follows the oldest supported Ubuntu LTS release.

HexF commented:
I'm struggling to understand why this test fails

It works fine locally..

phoerious commented:
Ancient Qt version, but it seems fine now. Only the test coverage is subpar and you need to rebase the branch.

yozoon commented:
I have a question regarding the _org.keepassxc.KeePassXC.policy_ file: Is the __auth_self_one_shot__ authorization still supported in newer Polkit versions? I couldn't find a reference to it in the man pages of the Polkit version shipped with Ubuntu 22.04. On my system (Arch Linux) Polkit even complains with the following message:
```polkit error: "Action org.keepassxc.KeePassXC.unlockDatabase is not registered"```
when I try to use fingerprint unlock (and fails to present the Polkit dialog) unless I change the action from __auth_self_one_shot__ to __auth_self__. But I'm no Polkit expert, so maybe I simply overlooked something...

droidmonkey commented:
Agree, looks like only _auth_self_ and _auth_self_keep_ are valid user space options.

droidmonkey commented on share/linux/org.keepassxc.KeePassXC.policy.in:
```diff
@@ -0,0 +1,18 @@
+    <message>Authentication is required to unlock a KeePassXC Database</message>
+    <defaults>
+      <allow_inactive>no</allow_inactive>
+      <allow_active>auth_self_one_shot</allow_active>
```
This needs to be changed to just _auth_self_

HexF replied:
Yep - that seems to have changed in polkit since I wrote the policy

HexF commented:
Rebased on to current develop branch

droidmonkey commented:
Hopefully can get this merged this weekend. This is a great change list and allows me to easily implement pin unlock as well.

Pacman99 commented:
~~How did you get this working on NixOS? I was able to override keepassxc with your commit and keyutils build input to get it running. But after I lock my database and then click on the Quick Unlock button I get the error `Failed to authenticate with Quick Unlock` on the red banner above. I see you mentioned something about adding a polkit action and I'm not sure how to do that. Thanks for working on this feature, I would love to try it out!~~

Figured it out! I needed to add keepassxc to system packages and not user/home packages to make sure the policykit action got loaded. I can confirm that this works great on nixos 23.05

droidmonkey commented:
Need to document that fact. @Pacman99 what did you do exactly? I am not familiar with what you mentioned here.

HexF commented:
This issue is to be expected when keepassxc is not installed in the system and hence the polkit action isn't registered with polkit.

Pacman99 commented:
Maybe a similar problem could occur in other distros if your were to install keepassxc through flatpak, appimages, or in your home directory and ran from there. So could be worth it to note that for quick unlock keepassxc has to be installed systemwide through the official package manager.

Or could there be a better error statement if the polkit action hasn't been registered?

HexF commented:
QuickUnlock via Polkit could be omitted from compilation for packaging formats such as Flatpak to avoid this issue.
If you check the stderr, you should see a polkit error message which arises when the action is not registed. It would be best to forward this to the UI.

HexF commented:
Is there anything holding up merging this PR?

HexF committed:
Add Polkit Quick Unlock Support

Closes #5991
Closes #3337 - Support fingerprint readers on Linux

Polkit allows for authentication of many means, including fingerprint scanning. Furthermore, a common interface for Quick Unlocking has been implemented, and has been replaced throughout to make implementing other quick unlock strategies easier.

Refactor QuickUnlock to use UUID stored in headers. This is a new feature using the KDBX 4 standard to store a randomly generated UUID in the public headers of the database. This enables identification of KDBX file without relying on path or filename and will eventually support persistent Quick Unlock.

spikespaz commented:
For me, on 2.7.6 this patch fails to apply at hunk 6. It's going to need an update before merge.

HexF commented:
Anything blocking merging?

droidmonkey commented:
Yes, I need to cleanup the code a little more (style/format issues).

I also realized that your implementation of the common class only allows one type of Quick Unlock at a time. This is fine for now, but in order to introduce pin-based quick unlock we would need to allow multiple options. I was thinking of shifting to a registration based class where there is a QuickUnlockManager and it would have an array of quick unlock options associated with it (depending on platform). Alternatively, we could always instantiate up to two QuickUnlock options, pin-based and platform specific.

What are your thoughts?

HexF commented:
I'm leaning more towards a QuickUnlockManager long term, but for now where we only have a single implementation per platform it doesn't make sense to implement.
The only real refactoring to implement such a manager would be changing the hooking - implementations could remain the same.

rgarcia89 commented:
Where are we with this PR? I can't see it anymore how my colleagues unlock KeePassXC on their Macs with the fingerprint every time and I'm left with nothing but the password input. :laughing:

droidmonkey commented:
It's basically ready we have a queue of merges to make

spikespaz commented:
Package maintainers eagerly await.

droidmonkey commented:
And done

northen-mindset commented:
> This issue is to be expected when keepassxc is not installed in the system and hence the polkit action isn't registered with polkit.

Appreciation for your efforts. I can't manage it to work. I've reinstalled it through `apt` instead of `flatpak`. Still there is no Quick Unlock functionality. Is there anything else one need to configure?

```
KeePassXC - Version 2.7.6
Revision: dd21def

Qt 5.15.3
Debugging mode is disabled.

Operating system: Ubuntu 22.04.3 LTS
CPU architecture: x86_64
Kernel: linux 6.2.0-37-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 2.19.1
```

droidmonkey commented:
Polkit quick unlock won't be available until 2.8.0 is released

HexF commented:
Do we have a rough date on when 2.8.0 will be released?

droidmonkey commented:
I'm hoping for shortly after the new year, we have some important changes in the pipeline right now

RobieWoods commented:
Hello @HexF and @droidmonkey,

I have a question regarding this PR, @HexF, you wrote the following:
> The Polkit mechanism works similar to the existing TouchID implementation, by where the database key is encrypted using a 
random key which is stored in secure storage, in this case the [kernel-space keyring](https://man7.org/linux/man-pages/man2/add_key.2.html).

I would like to consider an [Evil maid attack](https://en.wikipedia.org/wiki/Evil_maid_attack) scenario. Let's suppose I have KeepassXC on my computer, my computer is locked but I'm not around. Is the Polkit Quick Unlock is as secure as typing the full password on Linux? If an attacker manage to dump the entire memory while I have left my device unattended, will he be capable to decipher the KeepassXC database if Quick Unlock is enabled even if my database is locked?

I have installed KeepassXC Git with Polkit integration, and it works great :tada:. I had the feeling at first that using this PR could increase both the security (since I can increase the password size and lock my database very quickly) and facilitate the ease of use. However, now I'm unsure about the security side. While reading this PR I had the feeling that the Quick Unlock using Polkit could allow an attacker to access the database even if it's locked by "simply" dumping the RAM of the device.

In any case, I'm glad this PR has been merged. I love using my fingerprint to decipher my Keepass database :sweat_smile:. So thank you!

Kind regards,
Rob

droidmonkey commented:
If you lose physical access to the machine such that a RAM dump can be accomplished, then not much can protect you. Also known as "its my computer now".

goetzc commented:
> If you lose physical access to the machine such that a RAM dump can be accomplished, then not much can protect you. Also known as "its my computer now".

What about encrypting that key (using the fingerprint data as the key for this) in user-space before adding it to the keyring, would this be possible?

BlackHoleFox commented:
Fingerprint sensors, by design, don't release their template data (the scan and representation of your fingerprint) outside of the scanner for security. OSes only get back a yes/no with some replay protection (depending what protocol its using).

droidmonkey commented:
Also fingerprints are FAR from random and would make terrible keys for encryption. The way polkit integration is implemented now is as secure as possible given the lack of kernel owned keyspace on Linux (i.e., the Linux kernel offers no way to handle sensitive data for user land). Other operating systems like macOS handle encryption keys on the kernel side (and via a hardware key store) to protect data from user land.

RobieWoods commented:
Okay, thanks for the reply!

In my setup I have a bash command that automatically lock all my databases when I press Meta+L before starting my locker. Could a command be designed to remove the temporary key in Polkit to avoid such attack?

I'm refering to this key:
https://github.com/keepassxreboot/keepassxc/blob/a8cfefe6c8a4c10f4a215a003a98148b56840474/src/quickunlock/Polkit.cpp#L190

HexF commented:
The key which is stored with the kernel is only one part in decoding the database master key, which is derived from more data stored in user-space.
For an evil-maid attack to be successful in this case, both kernel memory and user-space KeepassXC memory would need to be dumped, which may be possible using techniques such as ThunderSpy, but is unlikely to occur.
If you're that concerned about a memory dumping evil-maid attack, the best course of action would be to turn off the computer when left unattended.
If your just concerned about KeepassXC, you could try closing the application which would both remove the entry from the Linux keyring, and remove the other half of the key stored in user-space.

RobieWoods commented:
Actually, I was thinking that a dbus function could be created. Like the ones described in the [wiki](https://github.com/keepassxreboot/keepassxc/wiki/Using-DBus-with-KeePassXC). I would love to be able to do a:

```bash
$ qdbus org.keepassxc.KeePassXC.MainWindow /keepassxc org.keepassxc.KeePassXC.MainWindow.resetQuickUnlock
```

But even if KeepassXC doesn't have a Dbus function for that, I'm worried about the "kill the app" solution. The encryptedMasterKeys stored in the user space are only scrubbed once here in `Polkit::getKey`:
https://github.com/keepassxreboot/keepassxc/blob/a8cfefe6c8a4c10f4a215a003a98148b56840474/src/quickunlock/Polkit.cpp#L225


**I really think the encryptedMasterKeys should be scrubed in the two `Polkit::reset(...)` methods as well.** hashmap.remove() and hashmap.clear are not enough in my opinion. 
<details><summary>Click to see the links of the `Polkit::reset(...)` methods I'm refering to</summary>
https://github.com/keepassxreboot/keepassxc/blob/a8cfefe6c8a4c10f4a215a003a98148b56840474/src/quickunlock/Polkit.cpp#L99
https://github.com/keepassxreboot/keepassxc/blob/a8cfefe6c8a4c10f4a215a003a98148b56840474/src/quickunlock/Polkit.cpp#L84
</details> 

And finally, I think the destructor `Polkit::~Polkit()` should call the global Polkit::reset() to have everything scrubbed properly when the app is closed for good.

droidmonkey commented:
We scrub all memory upon free.

RobieWoods commented:
> We scrub all memory upon free.

That's great, sorry, I didn't know that. I just read the operator overload of delete, it's very neat :)
I have two remaining questions regarding that subject:
1. This one is probably a noob one, since all data is scrubbed when being freed, is scrubbing the data on QByteArray really necessary [here](https://github.com/HexF/keepassxc/blob/73597e914f4ed827663c81ef19d775c39214a648/src/quickunlock/Polkit.cpp#L138) and [here](https://github.com/HexF/keepassxc/blob/73597e914f4ed827663c81ef19d775c39214a648/src/quickunlock/Polkit.cpp#L222)? Is it redundant? Or I forget something?
2. Would you guys be open to idea if I open a PR to add `org.keepassxc.KeePassXC.MainWindow.resetQuickUnlock` dbus option?

droidmonkey commented:
The reset quick unlock dbus command is super specific to your use case / desires. I personally don't see purpose in something like that and truly defeats the purpose of quick unlock in the first place. It seems like you want to, but not really, use quick unlock. 

Those memory scrubs are overkill, but sometimes qt can be a little clingy to containers like QString and QByteArray so it doesn't hurt to explicitly clear the buffer while you have hold of it.

RobieWoods commented:
Okay thanks! I think you are right. My request is too specific and can be handled by a simple script (at least for my case since KeepassXC always start in background). I just needed to call a small script before calling my Locker.

If that script interest some people, here it is:

```bash
#!/usr/bin/env bash

function back() {
  nohup $@ > /dev/null 2>&1 &
}

KEEPASSXC_DBUS_STR="org.keepassxc.KeePassXC.MainWindow"

if timeout 5 qdbus ${KEEPASSXC_DBUS_STR} /keepassxc lockAllDatabases ; then
  # exit keepassxc and restart it to reset the QuickUnlock keys
  # see: https://github.com/keepassxreboot/keepassxc/pull/8983#issuecomment-1876248571
  qdbus ${KEEPASSXC_DBUS_STR} /keepassxc appExit && back keepassxc
else
  echo "Warning: Cannot Lock KeepassXC database"
fi
```

rgarcia89 commented:
Someone from the maintainer team knows if an early release including this feature is possible? I am seeing a lot of still open backlog items for 2.8.0. 

This feature seems to be very awaited by the Linux users.

PS: I am aware of the possibility to build it by myself 😉

Arbel-arad commented:
i tried using the polkit authentication on arch, but it's giving me some errors:
`nixutils: failed to parse  "/proc/3/stat"`
![image](https://github.com/keepassxreboot/keepassxc/assets/65590498/3703d5ba-d3df-44e5-bdbf-f90bc39c9082)
```
KeePassXC - Version 2.8.0-snapshot
Build Type: Snapshot
Revision: a472ef8

Qt 5.15.12
Debugging mode is disabled.

Operating system: Arch Linux
CPU architecture: x86_64
Kernel: linux 6.7.6-zen1-1-zen

Enabled extensions:
- Auto-Type
- Browser Integration
- Passkeys
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 3.3.0
```

droidmonkey commented:
Ahh yes, this is important: nixutils: failed to parse "/proc/3/stat"

Can you share the contents of that file? It's the process information for keepassxc.

rgarcia89 commented:
I am seeing the same on manjaro linux.

```
nixutils: failed to parse  "/proc/351389/stat"
```

content of the file
```
351389 (keepassxc) S 351236 344368 344368 0 -1 4194304 20684 691 339 0 283 37 0 0 20 0 7 0 20803161 1217449984 55042 18446744073709551615 1 1 0 0 0 0 0 16781312 16391 0 0 0 17 3 0 0 0 0 0 0 0 0 0 0 0 0 0
```

droidmonkey commented:
I found the error! #10350

rgarcia89 commented:
Awesome I look forward to the snapshot build

rgarcia89 commented:
@droidmonkey still not working with the latest snapshot build.

Getting the same error

![image](https://github.com/keepassxreboot/keepassxc/assets/4134730/9bc11b2e-ee75-494a-b501-6f3261aec335)


```
 nixutils: failed to parse  "/proc/494940/stat"
```

content of the file
```
494940 (keepassxc) S 494754 1945 1945 0 -1 4194304 17338 685 336 0 77 8 0 0 20 0 8 0 55832569 1272360960 38561 18446744073709551615 1 1 0 0 0 0 0 16781312 16391 0 0 0 17 4 0 0 0 0 0 0 0 0 0 0 0 0 0
```

droidmonkey commented:
It's not merged yet.....

rgarcia89 commented:
true story - totally missed that
